File: server/nodejs/src/hostedcheckout/hostedCheckoutController.js
const config = require('../configuration/config')
const mapper = require('./hostedCheckoutMapper')
const hostedCheckoutService = require('./hostedCheckoutService')
const paymentDetailsService = require('../payment/paymentDetailsService')

/**
 * Initializes hosted checkout request
 */
exports.initializeRequest = () => {
    return mapper.toEmptyDto();
}

/**
 * Gets a payment response for a provided hosted checkout id as a query param
 */
exports.getPaymentResponse = async function (req) {
    if (req.query.hostedCheckoutId === undefined || req.query.hostedCheckoutId === '')
        throw new Exception("Hosted checkout id is missing");

    const hostedCheckoutId = req.query.hostedCheckoutId;    

    const paymentDetailsResponse = await paymentDetailsService.getPaymentDetailsForHostedCheckout(hostedCheckoutId);
    return paymentDetailsResponse.body
}

/**
 * Performs a hosted checkout request
 */
exports.createHostedCheckout = async function (req) {
    const hostedCheckoutRequest = mapper.toCreateHostedCheckoutRequest(req.body);
    const hostedCheckoutResponse = await hostedCheckoutService.createHostedCheckoutResponse(hostedCheckoutRequest)
    return hostedCheckoutResponse.body
}

File: server/dotnet/HostedCheckout/HostedCheckoutController.cs
using System.Runtime.ConstrainedExecution;
using Microsoft.AspNetCore.Mvc;
using OnlinePayments.Sdk.Domain;

namespace OnlinePayments.Sdk.Example;

[Route("api/hostedcheckout")]
[ApiController]
public class HostedCheckoutController : ControllerBase
{
    private readonly HostedCheckoutMapper _hostedCheckoutMapper;
    private readonly HostedCheckoutService _hostedCheckoutService;
    private readonly PaymentDetailsService _paymentDetailsService;

    public HostedCheckoutController(
        HostedCheckoutMapper hostedCheckoutMapper,
        HostedCheckoutService hostedCheckoutService,
        PaymentDetailsService paymentDetailsService
    )
    {
        _hostedCheckoutMapper = hostedCheckoutMapper;
        _hostedCheckoutService = hostedCheckoutService;
        _paymentDetailsService = paymentDetailsService;
    }
    
    [HttpGet]
    public CreateHostedCheckoutBasicDto GetHostedCheckout() {
        return _hostedCheckoutMapper.ToEmptyDto();
    }

    [HttpGet("outcome")]
    public async Task<PaymentDetailsResponse> GetPaymentResponse(
        [FromQuery] string RETURNMAC,
        [FromQuery] string hostedCheckoutId
    ) {
        if (hostedCheckoutId == null)
            throw new Exception("Hosted checkout id is missing!");

        return await _paymentDetailsService.GetPaymentDetailsForHostedCheckout(hostedCheckoutId);

    }

    [HttpPost("basic")]
    public async Task<CreateHostedCheckoutResponse> CreateHostedCheckout(
        [FromBody] CreateHostedCheckoutBasicDto createHostedCheckoutBasicDto
    ) {
        return await _hostedCheckoutService.CreateHostedCheckoutResponse(
                _hostedCheckoutMapper.ToCreateHostedCheckoutRequest(createHostedCheckoutBasicDto));
    }

}
